Telescope scopus

tuanchien opened the pull request:
Design doc for the SCOPUS source.

Particular points of interest:
 - naming conventions
 - the use of dynamic dags
 - whether to use dynamic tasks and partition the api retrieval task into separate tasks (either per call or per month or per year etc)
 - enforcement of similar bandwidth limits as wos
 - schema design. this is a stripped down version of the old framework. the old schema seemed to merge msa,wos,scopus + some other meta information. i've only put in scopus stuff as i didn't know what else you want in there.

tuanchien commented:
Thanks Rebecca.  Will do that when done.

Linking issue #20

aroelo reviewed (commented):
Nice work. I hope the existing telescopes can help you out a bit :-). 

I think the naming conventions for the connection & DAGs look good.

As a tip for the schema's what I sometimes do is upload the data to BigQuery using the console and 'auto detect' the schema. It then hopefully creates a new table from which I extract the schema. 
If you don't have any schema or a lot of missing fields in the one you do have this will usually help to get a basic schema which you can modify with extra fields/descriptions.

Please let me know if I can help with anything airflow/telescope related.

tuanchien commented on docs/telescopes/scopus.md:
````diff
@@ -0,0 +1,112 @@
+
+## DAG definition
+
+There will be a slight difference compared to the snapshot DAGs. It will query Airflow for a list of connection ids relevant to this particular data source, e.g., scopus_curtin, scopus_auckland, etc. A dynamically generated DAG will be used for each institution.
+
+The template for the dynamically generated DAGs will follow the pattern in the snapshot DAG, where the heavy lifting is done by Telescope classes.
+
+Listing airflow connections can be done with
+```python
+session.query(Connection)
+```
````
Thanks for your tips and suggestions Aniek.  I will aim to follow the style and architecture of the existing telescopes as much as possible.

I am fine with going the subdag route if you guys think this is a better fit for the project. Let me list my rationale for the current proposal though so you can evaluate my reasoning behind it.

A dynamic dag will show up as separate dags in the airflow interface.  This gives you visibility of the flow of execution for each task in the dag. The downside is if you have a lot of institutions, there will be lots of dags populated in the main airflow dag table.

The connection naming scheme allows the driver dag (the code that creates the separate (sub)dags) automatically discover institutions based on what connections have been configured in the airflow db with no code change.

SubDAGs currently hide the status of tasks in the airflow interface.  This has the benefit of neatly grouping the execution tasks for each institution under a single heading, and all of it under a main dag heading. The downside is we would have to rely on other mechanisms like logs for debugging.  It's possible this behaviour changes in future versions of airflow.

I have a preference for debug visibility, however, given the potential for many institutions, I also understand the desire to lump everything under a main dag heading.

jdddog replied:
These are the planned changes to subdags for Airflow 2.0: https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-34+TaskGroup%3A+A+UI+task+grouping+concept+as+an+alternative+to+SubDagOperator

They are on the next immediate steps: https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+2.0+-+Planning

tuanchien replied:
Let's go with subdags then, and migrate to taskgroups when 2.0 lands.

aroelo replied:
> SubDAGs currently hide the status of tasks in the airflow interface. This has the benefit of neatly grouping the execution tasks for each institution under a single heading, and all of it under a main dag heading. The downside is we would have to rely on other mechanisms like logs for debugging. It's possible this behaviour changes in future versions of airflow.

Haven't tried it myself, but according to the documentation you can zoom in on SubDags and view the tasks. Not sure if you meant this interface?

tuanchien replied:
Ah good spotting.  Yeah I meant that one.  I haven't tried it either, but was going off what I thought the udemy course said.

jdddog commented on docs/telescopes/scopus.md:
```diff
@@ -0,0 +1,112 @@
+
+## download_data
+
+From the start date til the most recently completed month, pull data from the data source using an adapted version of Richard’s code, and append to a file on GC containing a list of json responses. Not sure if you want timing information like you did before. Maybe also round robin multiple API keys when they exist for each query.
+
```
If we want to have a fresh copy of the most up to date data in Scopus for a particular institution, then we need to change the approach used to fetch data. The existing code only fetches data for a particular institution and month once, this means that fields such as the number of citations, will become out of date unless you fetch all of the records for each institution and date again, repeatedly.

So I think that we should change the architecture of how the telescope works to the following:
* The DAG runs every three months (could be a longer period, e.g. six months).
* Whenever the DAG runs, it fetches the records each institution, from a set date in the past (e.g. the year 2000) to the end date of the DAG run, even if it has fetched them before.
* All of the records fetched for each institution are given a release_date (the end date of the DAG run), which can be used to partition these records into a single release.

tuanchien commented:
Excuse the little bit of WosTelescope pollution.

Updates:
- ScopusTelescope follows design philosophy of WosTelescope
- Replaced elsapy with ScopusClient (custom)
- Using mocked out http endpoints rather than recording cassettes.
- Sequential download mode is round robin. [default]
- Parallel download mode has all keys fetch when they are ready.
- Download clients will sleep until they are ready if quota timed out, i.e., policy of no failure due to quota exceeding.
- Support downloading multiple institutions with single query
- Documentation update [both wos and scopus] and provide extra links to API reference docs.
- Implement SchedulePeriod instead of using tuple for periods
- Added ScheduleOptimiser solver for minimising api calls (not connected to any telescopes)
- Add 'view' as a parameter for each api key to allow different access levels.
- Added support for parsing extra fields if key has 'complete' view access.  Not live tested since Curtin key doesn't have this access level.
- Added BQ schema
- Minor wos test / telescope changes to reflect some of the architecture tweaks. [will make a separate pr for cutting cassette out of wos tests since it's a bigger job]

jdddog reviewed (commented):
This is looking really good Tuan, I just have a couple of suggestions:

* The transformed files are currently stored in the downloads folder, it would be good to put them into the transformed folder, to make it consistent with the other telescopes, it also makes it easy to find the different types of files.
* It would also be good to put the files for a particular institute into their own folders, similar to what you did for the WoS telescope, so that it is easy to identify files for a particular institute.

tuanchien commented:
Updated:
- Using pendulum.Period instead of SchedulePeriod (removed)
- Fixed download/transform local paths.
- Make author and affiliations fields return None if empty list

tuanchien commented:
Updated to use standardised user agent string from get_ao_user_agent()

jdddog reviewed (approved):
Looks good Tuan!
